subsys/bluetooth/services/dis.c

```diff
@@ -101,6 +138,75 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
 }
 #endif
 
+#if CONFIG_BT_DIS_SYSTEM_ID
+static ssize_t read_system_id(struct bt_conn *conn,
+			   const struct bt_gatt_attr *attr, void *buf,
+			   uint16_t len, uint16_t offset)
+{
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, dis_system_id,
+				 sizeof(dis_system_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_UDI
+#define BT_DIS_UDI_FLAG_LABEL		(!!BT_DIS_UDI_LABEL_STR_REF[0])
+#define BT_DIS_UDI_FLAG_DI			(!!BT_DIS_UDI_DI_STR_REF[0])
+#define BT_DIS_UDI_FLAG_ISSUER		(!!BT_DIS_UDI_ISSUER_STR_REF[0])
+#define BT_DIS_UDI_FLAG_AUTHORITY	(!!BT_DIS_UDI_AUTHORITY_STR_REF[0])
+#define BT_DIS_UDI_FLAGS			(BT_DIS_UDI_FLAG_LABEL | (BT_DIS_UDI_FLAG_DI << 1) | \
+	(BT_DIS_UDI_FLAG_ISSUER << 2) | (BT_DIS_UDI_FLAG_AUTHORITY << 3))
+
+#if defined(CONFIG_BT_SETTINGS)
+static uint8_t bt_dis_udi_merged[MIN(CONFIG_BT_DIS_STR_MAX * 4 + 1,
+	MIN(BT_ATT_MAX_ATTRIBUTE_LEN, CONFIG_BT_L2CAP_TX_MTU))];
+#else
+static uint8_t bt_dis_udi_merged[MIN(BT_ATT_MAX_ATTRIBUTE_LEN, CONFIG_BT_L2CAP_TX_MTU)];
+#endif
+
+static size_t bt_dis_udi_merged_size = 0;
+
+static ssize_t read_udi(struct bt_conn *conn,
+			   const struct bt_gatt_attr *attr, void *buf,
+			   uint16_t len, uint16_t offset)
+{
+	if (!bt_dis_udi_merged_size) {
```

Comment:
Is it necessary to cache the merged value? We can construct the response in `buf` directly on-the-fly, instead of using `bt_gatt_attr_read`.

Reply:
Perhaps this is the simplest solution for `offset`. If I understand it right, this is used for long reads, where one only needs part of the characteristic? It might be simpler to just construct the entire thing and then index, since we do not know ahead of time how long each part is.
The caching is just to save work if called multiple times, but it could be removed to save memory. The buffer could just live on the stack.

Reply:
This is a place where the GATT API is a bit weird to use. `bt_gatt_attr_read` expects a value and an offset, which means that in order to use that, the provided value shall contain the entire final result, after which the offset is then applied to copy the value into the buffer. 

This function can be optimized to construct the long value only partially given the offset and call `bt_gatt_attr_read` with offset=0 regardless of the actual offset, which I think is what @alwa-nordic is suggesting. 

It is a significant memory optimization, as we do not need to allocate for the strings twice (saving up to 512 octets).

Reply:
We can easily construct an offset read result if we define a helper function. Something like this:

```c
/* Emulate the three step process, with an infinitely sized tmp:
 *  - memcpy(&tmp[virt_dst], dst, dst_size); // Initialize tmp
 *  - memcpy(&tmp[virt_src], src, src_size); // Overwrite part
 *  - memcpy(dst, &tmp[virt_dst], dst_size); // Copy back
 */
void memcpy_clip(void *dst, size_t dst_size, size_t virt_dst, size_t virt_src, void *src, size_t src_size)
{
    size_t virt_copy_start = MAX(virt_src, virt_dst);
    size_t virt_copy_end = MIN(virt_src + src_size, virt_dst + dst_size);
    if (virt_copy_start < virt_copy_end)
    {
        size_t cpy_size = virt_copy_end - virt_copy_start;
        (void)memcpy(dst, src, cpy_size);
    }
}

void usage(size_t read_buf, size_t read_size, size_t read_offset)
{
    size_t virt_src = 0;
    memcpy_clip(read_buf, read_size, read_offset, virt_src, field1, field1_size);
    virt_src += field1_size;
    memcpy_clip(read_buf, read_size, read_offset, virt_src, field2, field2_size);
    ...
}
```

Reply:
I implemented a way to find the correct string for an index, so that the value can be constructed directly into the buffer. Also, I verified that this works with long reads.